zfs_2_1: init at 2.1.13

RaitoBezarius opened the pull request:
## Description of changes

This re-introduces the old "super" (in the sense more stable than the average stable ZFS version.) stable ZFS version we had in the past following the many predicted issues of ZFS 2.2.x series.

I am also removing myself from maintenance of any further ZFS versions as I am planning to quit ZFS maintenance at some point.

In the meantime, for users like me who depend on ZFS for critical operations, here is a ZFS version that is known to work for LTS kernels.

I invite anyone who care about ZFS to mention themselves in this issue or to add themselves as a maintainer.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

### Priorities

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RaitoBezarius committed:
zfs_2_1: init at 2.1.13

This re-introduces the old stable ZFS version we had in the past following
the many predicted issues of ZFS 2.2.x series, that is much more stable
than any further ZFS version at the moment.

I am also removing myself from maintenance of any further ZFS versions as I am
planning to quit ZFS maintenance at some point.

In the meantime, for users like me who depend on ZFS for critical operations, here is a ZFS version
that is known to work for LTS kernels.

RaitoBezarius commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -210,13 +210,21 @@ in
       };
 
+      modulePackage = mkOption {
+        type = types.package;
```
This is necessary to enable users to choose their "kernel module" package.

RaitoBezarius replied:
Maybe the best way to go about it is to implement a passthru that reimports the zfs derivation in kernel mode and just reuse the passthru to avoid any possible divergence.

Mic92 replied:
I don't quite understand why you want different kernel and userspace zfs versions?

Mic92 replied:
I am wondering if we should just drop the enableUnstable flag eventually and just have this option?

RaitoBezarius replied:
It sounds good to me, but do we have any situation where the ZFS package and module package can be meaningfully different?

RaitoBezarius replied:
I can make it internal and do the smart select inside the NixOS module.

thiagokokada commented on nixos/tests/zfs.nix:
```diff
@@ -193,6 +196,10 @@ let
 
 in {
 
+  superstable = makeZfsTest "superstable" {
+    zfsSuffix = "SuperStable";
+  };
```
Maybe `oldstable` would be better? Because what exactly `superstable` means without context?

RaitoBezarius replied:
It has a better "stability" property than the average property (etymology of `super`). I think it's more accurate to describe what it is than `oldstable` given that oldstable and stable doesn't have the same stability quality.

RaitoBezarius replied:
I can call it `zfsThatDoesNotEatYourData` if you prefer.

thiagokokada replied:
> It has a better "stability" property than the average property

But is this something that upstream gives? Otherwise we are just inventing names?

We could call this `oldstable` (because at least it is descriptive of what it does) or something else and add something like this to the `meta.longDescription`:

```
This is the old stable version known to work with LTS kernels and avoids some of the possible data corruption issues that may be introduced in more recent releases.
```

But that is just my 2 cents, I don't use ZFS at all.

RaitoBezarius replied:
Well, naming is a privilege of the people putting the work in fixing things. ZFS has proved they are unable to deal with predicted issues, so I think they lost their naming privilege here.

For the long description, I agree.

For old stable, I honestly don't like it again because it doesn't reflect the difference in quality. This can change when ZFS would prove itself again.

adamcstephens replied:
I don't love `superStable` either. If we're going to be supporting 2.1 long term, would it make more sense to put the versions in the attribute name? e.g. `zfs_2_1` as we do elsewhere?

RaitoBezarius replied:
OK, I can do that

RaitoBezarius replied:
I'm not changing to `zfsStable` to `zfs_2_1` FWIW BTW. That's the responsibility of the new maintainers of ZFS.

RaitoBezarius replied:
(someone has to figure out if 2.2 to 2.1 is possible, etc. I don't want to spend that time personally.)

RaitoBezarius replied:
I suppose the naming discussion is now finished, resolving, let me know if you still have concerns.

RaitoBezarius commented:
> @RaitoBezarius I’m willing to be added as a zfs maintainer if wanted, though I admit I’m not yet deeply familiar with the nixpkgs drv and test code.

I added you as a maintainer. Well, I was not deeply familiar with anything too when I entered into the ZFS maintenance. :)

RaitoBezarius commented:
I added @adamcstephens as a default maintainer too, let me know if you want me to remove it.

RaitoBezarius committed:
zfs: improve description and long description

When we are building for user or kernel, we are not building a kernel module all the time.

RaitoBezarius commented:
```
/nix/store/bzk7xh8c84hxj1kiwgcpm4kn96y2g5bq-vm-test-run-multi-disk-zfs
/nix/store/v8f3r8f19gmpbsmcarf1nvis6sddrwvg-vm-test-run-installer-zfs-root
/nix/store/50z39myqynglp36z9gf9sqv5fs3iyrx8-vm-test-run-zfs-2.1-series
/nix/store/3zj6r06w6a4whz3r8mplphnbwpcz9zpb-vm-test-run-zfs-stable
/nix/store/bl5qy909v1jc6snd12j65zm45zk6z63z-vm-test-run-zfs-unstable
/nix/store/1fsi8pxi87m2507zppqki6sfmmlmnb8n-vm-test-run-zfs-unstable
```

RaitoBezarius commented:
This is ready for review @Mic92 @adamcstephens @amarshall, PTAL.

SuperSandro2000 commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -203,18 +204,26 @@ in
 
   imports = [
     (mkRemovedOptionModule [ "boot" "zfs" "enableLegacyCrypto" ] "The corresponding package was removed from nixpkgs.")
+    # Introduce me in 24.05.
+    # (mkRemovedOptionModule [ "boot" "zfs" "enableUnstable" ] "`boot.zfs.enableUnstable` was replaced by `boot.zfs.package` where you can specify `pkgs.zfsUnstable` to track the unstable ZFS.")
```
Can we put this into an extra commit and not backport it?

RaitoBezarius replied:
Actually, we should use 24.11.

SuperSandro2000 commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -203,18 +204,26 @@ in
-        defaultText = literalExpression "if config.boot.zfs.enableUnstable then pkgs.zfsUnstable else pkgs.zfs";
-        description = lib.mdDoc "Configured ZFS userland tools package.";
+        default = if cfgZfs.enableUnstable then pkgs.zfsUnstable else pkgs.zfs;
+        defaultText = literalExpression "if unstable is enabled then pkgs.zfsUnstable else pkgs.zfs";
```
```suggestion
        defaultText = literalExpression "if zfsUnstable is enabled then pkgs.zfsUnstable else pkgs.zfs";
```
otherwise people might confuse this with nixos-unstable

SuperSandro2000 commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -203,18 +204,26 @@ in
-        description = lib.mdDoc "Configured ZFS userland tools package.";
+        default = if cfgZfs.enableUnstable then pkgs.zfsUnstable else pkgs.zfs;
+        defaultText = literalExpression "if unstable is enabled then pkgs.zfsUnstable else pkgs.zfs";
+        description = lib.mdDoc "Configured ZFS userland tools package, use `pkgs.zfsUnstable` if you want to track the unstable ZFS.";
```
```suggestion
        description = lib.mdDoc "Configured ZFS userland tools package, use `pkgs.zfsUnstable` if you want to track the latest ZFS.";
```

RaitoBezarius replied:
Nope, this is not the latest ZFS. This is the staging ZFS technically.

SuperSandro2000 commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -203,18 +204,26 @@ in
+
+      modulePackage = mkOption {
+        internal = true; # It is supposed to be selected automatically, but can be overridden by expert users.
+        default = (selectModulePackage cfgZfs.package);
```
```suggestion
        default = selectModulePackage cfgZfs.package;
```

SuperSandro2000 commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -533,6 +542,13 @@ in
       assertions = [
+        {
+          assertion = cfgZfs.modulePackage.version == cfgZfs.package.version;
+          message = ''
+            The kernel module and the userspace tooling versions
+            are not matching, this is an unsupported usecase.
+          '';
```
```suggestion
          message = "The kernel module and the userspace tooling versions are not matching, this is an unsupported usecase.";
```
not need to span this over two lines in the terminal

SuperSandro2000 commented on nixos/modules/tasks/filesystems/zfs.nix:
```diff
@@ -571,18 +587,14 @@ in
       boot.initrd = mkIf inInitrd {
-        kernelModules = [ "zfs" ] ++ optional (!cfgZfs.enableUnstable) "spl";
+        # spl has been removed in ≥ 2.2.0.
+        kernelModules = [ "zfs" ] ++ lib.optional (lib.versionAtLeast version "2.2.0") "spl";
```
```suggestion
        kernelModules = [ "zfs" ] ++ lib.optional (lib.versionOlder "2.2.0" version) "spl";
```
I was really confused about this. According to the docs the first argument is what we are checking for and the second the variable which also makes in my brain much more sense https://github.com/NixOS/nixpkgs/blob/master/lib/strings.nix#L686-L706

Also before the condition was intuitively not matching the comment which it is now.

SuperSandro2000 commented on pkgs/os-specific/linux/zfs/2_1.nix:
```diff
@@ -0,0 +1,40 @@
+  stdenv' = if kernel == null then stdenv else kernel.stdenv;
+in
+callPackage ./generic.nix args {
+  # You have to ensure that in `pkgs/top-level/linux-kernels.nix`
+  # this attribute is the correct one for this package.
```
There is not a free way to assert this? Might be over engineered.

RaitoBezarius replied:
Not really, I leave this open on purpose because you might have a reason to use an invalid attribute locally for testing.

SuperSandro2000 commented on pkgs/os-specific/linux/zfs/2_1.nix:
```diff
@@ -0,0 +1,40 @@
+  # if the 2.1.x series moves.
+  version = "2.1.13";
+
+  sha256 = "sha256-tqUCn/Hf/eEmyWRQthWQdmTJK2sDspnHiiEfn9rz2Kc=";
```
```suggestion
  hash = "sha256-tqUCn/Hf/eEmyWRQthWQdmTJK2sDspnHiiEfn9rz2Kc=";
```
if we are completely reworking generic.nix we might as well.

SuperSandro2000 commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -1,222 +1,242 @@
+
+    buildKernel = any (n: n == configFile) [ "kernel" "all" ];
+    buildUser = any (n: n == configFile) [ "user" "all" ];
+    is22Series = versionAtLeast version "2.2.0";
```
```suggestion
    isAtLeast22Series = versionAtLeast version "2.2.0";
```

Since this is only used in two conditions below, we might as well use if then else to make the intent clear.

SuperSandro2000 commented on pkgs/os-specific/linux/zfs/generic.nix:
```diff
@@ -1,222 +1,242 @@
+  in
+
+  stdenv'.mkDerivation {
+    name = "zfs-${configFile}-${version}${optionalString buildKernel "-${kernel.version}"}";
+    inherit version;
```
If we set name and version, shouldn't we also set pname?

SuperSandro2000 reviewed (commented):
just some minor details, otherwise LGTM. Also the name is now way better 😉

ius commented:
Note we might not be completely out of the woods with zfs v2.1 as far as corruption issues go, as it's _also_ affected (see https://github.com/openzfs/zfs/issues/15533). There's a (very new) PR allegedly fixing it.

Best summary of all open corruption issues I could find: https://github.com/openzfs/zfs/issues/15526#issuecomment-1825113314

And I've completely lost track as to what workaround apply to which zfs version..

RaitoBezarius commented:
> Note we might not be completely out of the woods with zfs v2.1 as far as corruption issues go, as it's _also_ affected (see [openzfs/zfs#15533](https://github.com/openzfs/zfs/issues/15533)). There's a (very new) PR allegedly fixing it.
> 
> Best summary of recent open corruption issues I could find: [openzfs/zfs#15526 (comment)](https://github.com/openzfs/zfs/issues/15526#issuecomment-1825113314)
> 
> And I've completely lost track as to what workaround applies to which zfs version..

I pulled the fix in the upcoming push I am going to do.

RaitoBezarius commented:
Merging because this is a critical PR for people who depend on non-corrupted filesystem, we can perform many fixups after this one, I will not merge the backport before we discuss it again.
